helpers/embeds.js

```diff
@@ -0,0 +1,75 @@
+import { html, render } from 'lit';
+import { findComposedAncestor } from '../helpers/dom.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		await Promise.all(this.#childNodes.map(childNode => childNode.renderNode(options)));
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) {
+			render(embed, this.#node.parentNode);
```

Comment:
I just jammed this in as a quick "Would Lit work here!?" test, and yeah, this works fine. In practice, if we wanted to do this using Lit, I'm sure there's a more clever way we could manage this to avoid direct DOM manipulation, at least for the html-block. In Content I think this is always going to be our lot in life.

Reply:
Interesting. What value does using Lit's `render` provide here? Just making the rendering `async`, with the hope all the replacements would pop in at the same time? Note that we're already directly/synchronously updating the DOM with the following `node.remove()` statement.

Reply:
For Content files in particular, nothing really. I see it being a bigger advantage in the case where this gets used within the html-block, since then this code could return a template that the html-block could insert in its own `render` method. Of course a potential benefit on the plugin side of things is that more complex structures could be more easily constructed via Lit templates. Otherwise plugins would need to build out their own DOM elements, which I imagine could get complex and annoying fast for complex components.